File: 백준/Silver/1743. 음식물 피하기/음식물 피하기.py
import sys
from collections import deque
input = sys.stdin.readline

n, m, k = map(int, input().split())

graph = [[0] * m for _ in range(n)]
visited = [[0] * m for _ in range(n)]

for _ in range(k):
    r, c = map(int, input().split())
    graph[r - 1][c - 1] = 1

# def dfs(x, y, cnt):
#     visited[x][y] = cnt

#     dx = [-1, 1, 0, 0]
#     dy = [0, 0, -1, 1]

#     for i in range(4):
#         nx = x + dx[i]
#         ny = y + dy[i]

#         if 0 <= nx < n and 0 <= ny < m and graph[nx][ny] == 1 and not visited[nx][ny]:
#             dfs(nx, ny, cnt + 1)

def bfs(r, c):
    q = deque([(r, c)])
    visited[r][c] = 1

    cnt = 1

    while q:
        x, y = q.popleft()

        dx = [-1, 1, 0, 0]
        dy = [0, 0, -1, 1]

        for i in range(4):
            nx = x + dx[i]
            ny = y + dy[i]

            if 0 <= nx < n and 0 <= ny < m and graph[nx][ny] == 1 and not visited[nx][ny]:
                cnt += 1
                visited[nx][ny] = cnt
                q.append((nx, ny))

for i in range(n):
    for j in range(m):
        if graph[i][j] == 1 and not visited[i][j]:
            bfs(i, j)

# for i in graph:
#     print(i)

# print()

ans = 0
for i in visited:
   for j in i:
       if j > ans:
           ans = j
print(ans)

File: 백준/Silver/1743. 음식물 피하기/음식물 피하기.cc
#include <iostream>
#include <queue>
using namespace std;

int N, M, K;
int graph[100][100] = { 0 };
int visited[100][100] = { 0 };
queue<pair<int, int>> q;

void bfs(int r, int c) {
    q.push({r, c});
    visited[r][c] = 1;

    int cnt = 1;

    while (!q.empty()) {
        int x = q.front().first;
        int y = q.front().second;
        q.pop();

        int dx[4] = {-1, 1, 0, 0};
        int dy[4] = {0, 0, -1, 1};

        for (int i = 0; i < 4; i++) {
            int nx = x + dx[i];
            int ny = y + dy[i];

            if (nx >= 0 and nx < N and ny >= 0 and ny < M) {
                if (graph[nx][ny] == 1 && !visited[nx][ny]) {
                    cnt++;
                    visited[nx][ny] = cnt;
                    q.push({nx, ny});
                }
            }
        }
    }
}

int main() {
    cin >> N >> M >> K;

    for (int i = 0; i < K; i++) {
        int r, c;
        cin >> r >> c;
        graph[r - 1][c - 1] = 1;
    }


    for (int i = 0; i < N; i++){
        for (int j = 0; j < M; j++) {
            if (graph[i][j] == 1 && !visited[i][j]) {
                bfs(i, j);
            }
        }
    }

    int ans = 0;

    for (int i = 0; i < N; i++){
        for (int j = 0; j < M; j++) {
            if (visited[i][j] > ans) {
                ans = visited[i][j];
            }
        }
    }

    cout << ans;
    
    return 0;
}